### PR TITLE
Use getrfBatched in `linalg.slogdet`

### PR DESCRIPTION
**Pull request by toslunar**
`linalg.det` using `getrfBatched` is originally implemented in chainer/chainer#613.

For the input `random.randn(256, 256, 256)`,

- v6: 0.85 sec
- v7 (after #2437 and #2660): 1.2 sec
- this PR: 0.022 sec

### CODE REVIEW

**Comment by toslunar** [comment]
`getrfBatched` can be much slower than `getrf`.

`linalg.det(random.randn(4096, 4096).astype(float32))`
- v7: 0.069 sec
- this PR: 11.29 sec

**Comment by emcastillo** [comment]
This may not make any sense at all but I am just throwing the ball here.

Could we measure how long does it take to do it using cusolver QR decomposition instead of cublas batched LU?

**Comment by emcastillo** [comment]
https://devtalk.nvidia.com/default/topic/902471/cublas-very-low-occupancy/

>> cublas getrfBatched and getriBatched are not intended to be high performance for a single large matrix. Their intended use is for a large number of "small" matrices, where the kernel launch overhead would be a significant contributor to the overall execution time, if they were inverted one-by-one.

**Comment by toslunar** [comment]
I compared the speed of `linalg.det` for input size `(2**i, 2**j, 2**j)`.
```
log(time_batched / time_not_batched)

i\j   0      1      2      3      4      5      6      7      8      9     10     11     12
 0 [[ 0.059 -0.064 -0.091 -0.084 -0.058 -0.123 -0.159 -1.025 -0.225  0.923  2.223  3.795  5.099]
 1  [-0.151 -0.15  -0.109 -0.158 -0.112 -0.23  -1.412 -1.585 -0.821  0.287  1.696  3.159  4.455]
 2  [-0.314 -0.297 -0.251 -0.29  -0.269 -1.384 -1.994 -2.217 -1.481 -0.376  1.063  2.486  3.773]
 3  [-0.538 -0.516 -0.488 -0.565 -1.191 -1.941 -2.614 -2.867 -2.155 -0.945  0.409  1.823    nan]
 4  [-0.847 -0.862 -0.808 -0.848 -1.772 -2.562 -3.226 -3.546 -2.791 -1.563 -0.241  1.219    nan]
 5  [-1.276 -1.305 -1.339 -1.391 -2.439 -3.231 -3.914 -4.165 -3.326 -2.146 -0.765  0.666    nan]
 6  [-1.874 -1.891 -1.889 -2.285 -3.079 -3.899 -4.619 -4.73  -3.673 -2.515 -1.25   0.111    nan]
 7  [-2.272 -2.306 -2.303 -2.793 -3.536 -4.379 -5.067 -4.997 -3.74  -2.737 -1.671    nan    nan]
 8  [-2.932 -2.872 -2.873 -3.449 -4.187 -5.043 -5.774 -4.841 -3.744 -2.889 -1.929    nan    nan]
 9  [-3.584 -3.598 -3.588 -4.131 -4.866 -5.741 -5.981 -4.825 -3.678 -2.857 -1.95     nan    nan]
10  [-4.267 -4.287 -4.56  -4.46  -5.492 -6.435 -6.012 -4.83  -3.681 -2.856 -1.839    nan    nan]
11  [-4.798 -4.768 -4.791 -5.482 -6.253 -7.1   -5.991 -4.834 -3.678 -2.842    nan    nan    nan]
12  [-5.608 -5.64  -5.612 -6.198 -6.936 -7.745 -5.98  -4.828 -3.663 -2.826    nan    nan    nan]]
```

**Comment by emcastillo** [comment]
These results seem to imply that for large few-stacked matrices getrfBatched can be way too slow.
And in some other cases way too fast.  I don't know if this trend holds on different gpus but having some heuristic to select which algorithm to use can have a big impact. 
Seems like in this case `#batches > 2**4` could be a good starting  point

**Comment by toslunar** [comment]
similar result for `float64`.
```
i\j   0      1      2      3      4      5      6      7      8      9     10     11     12
 0 [[-0.182 -0.088 -0.027  0.    -0.037 -0.074 -0.117 -1.029 -0.292  0.792  2.228  3.701  4.936]
 1  [-0.146 -0.124 -0.168 -0.139 -0.172 -0.221 -1.364 -1.601 -0.894  0.192  1.629  3.039  4.261]
 2  [-0.245 -0.254 -0.253 -0.272 -0.288 -1.356 -1.971 -2.223 -1.544 -0.329  0.973  2.364  3.6  ]
 3  [-0.416 -0.502 -0.469 -0.485 -1.114 -1.939 -2.592 -2.874 -2.185 -0.95   0.302  1.701  3.047]
 4  [-0.838 -0.842 -0.852 -0.865 -1.583 -2.567 -3.238 -3.527 -2.752 -1.595 -0.312  1.17     nan]
 5  [-1.262 -1.271 -1.286 -1.626 -2.285 -3.205 -3.91  -4.159 -3.34  -2.13  -0.822  0.531    nan]
 6  [-1.797 -1.827 -1.798 -2.301 -2.929 -3.854 -4.564 -4.623 -3.696 -2.524 -1.344 -0.039    nan]
 7  [-2.458 -2.441 -2.387 -2.942 -3.586 -4.53  -5.253 -4.9   -3.937 -2.926 -1.834    nan    nan]
 8  [-3.015 -3.04  -3.078 -3.603 -4.261 -5.238 -5.926 -4.731 -3.917 -3.033 -2.072    nan    nan]
 9  [-3.937 -3.707 -3.708 -4.301 -4.934 -5.916 -5.645 -4.474 -3.74  -2.832 -1.69     nan    nan]
10  [-4.389 -4.445 -4.671 -4.945 -5.638 -6.588 -5.777 -4.525 -3.772 -2.905    nan    nan    nan]
11  [-5.069 -5.076 -5.069 -5.662 -6.366 -7.264 -5.853 -4.549 -3.787 -2.919    nan    nan    nan]
12  [-6.117 -6.197 -6.101 -6.367 -7.065 -7.393 -5.905 -4.553 -3.786    nan    nan    nan    nan]]
```

File: cupy/linalg/decomposition.py [hunk 1]
```diff
@@ -7,6 +7,93 @@
...
+        if dtype == numpy.float32:
+            getrfBatched = cupy.cuda.cublas.sgetrfBatched
+        elif dtype == numpy.float64:
+            getrfBatched = cupy.cuda.cublas.dgetrfBatched
+        elif dtype == numpy.complex64:
+            getrfBatched = cupy.cuda.cublas.cgetrfBatched
+        elif dtype == numpy.complex128:
+            getrfBatched = cupy.cuda.cublas.zgetrfBatched
+        else:
+            assert False
```
Comment (emcastillo) [inline cupy/linalg/decomposition.py]: Can we rise a TypeError('Xx type is not supported')?

File: cupy/linalg/decomposition.py [hunk 2]
```diff
@@ -7,6 +7,93 @@
...
+            getrf = cusolver.sgetrf
+        elif dtype == numpy.float64:
+            getrf_bufferSize = cusolver.dgetrf_bufferSize
+            getrf = cusolver.dgetrf
+        elif dtype == numpy.complex64:
+            getrfBatched = cupy.cuda.cublas.cgetrfBatched
+        elif dtype == numpy.complex128:
+            getrfBatched = cupy.cuda.cublas.zgetrfBatched
+        else:
+            assert False
```
Comment (emcastillo) [inline cupy/linalg/decomposition.py]: Ditto

**Comment by toslunar** [comment]
The dtype check is in `slogdet`.  It is an option to raise an error in the private function `_lu_factor` instead, which seems to me
- less error-prone
- but not clean before `slogdet` supports complex types.

**Comment by emcastillo** [comment]
Ok! Thank you for clarification :)
The PR looks good to me then!

Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**